chia/util/keyring_wrapper.py

```diff
@@ -0,0 +1,349 @@
+import keyring as keyring_main
+
+from chia.util.default_root import DEFAULT_KEYS_ROOT_PATH
+from chia.util.file_keyring import FileKeyring
+from chia.util.misc import prompt_yes_no
+from keyrings.cryptfile.cryptfile import CryptFileKeyring  # pyright: reportMissingImports=false
+from pathlib import Path
+from sys import exit, platform
+from typing import Any, Optional, Tuple, Union
+
+
+# We want to protect the keyring, even if a user-specified master passphrase isn't provided
+DEFAULT_PASSPHRASE_IF_NO_MASTER_PASSPHRASE = "$ chia passphrase set # all the cool kids are doing it!"
+
+
+class KeyringWrapper:
+    """
+    KeyringWrapper provides an abstraction that the Keychain class can use
+    without requiring knowledge of the keyring backend. During initialization,
+    a keyring backend is selected based on the OS.
+
+    The wrapper is implemented as a singleton, as it may need to manage state
+    related to the master passphrase and handle migration from the legacy
+    CryptFileKeyring implementation.
+    """
+
+    # Static members
+    __shared_instance = None
+    __keys_root_path: Path = DEFAULT_KEYS_ROOT_PATH
+
+    # Instance members
+    keys_root_path: Path
+    keyring: Union[Any, FileKeyring] = None
+    cached_passphase: Optional[str] = DEFAULT_PASSPHRASE_IF_NO_MASTER_PASSPHRASE
+    cached_passphase_is_validated: bool = False
+    legacy_keyring = None
+
+    def __init__(self, keys_root_path: Path = DEFAULT_KEYS_ROOT_PATH):
+        """
+        Initializes the keyring backend based on the OS. For Linux, we previously
+        used CryptFileKeyring. We now use our own FileKeyring backend and migrate
+        the data from the legacy CryptFileKeyring (on write).
+        """
+        self.keys_root_path = keys_root_path
+        self.keyring = self._configure_backend()
+
+        # Configure the legacy keyring if keyring passphrases are supported to support migration (if necessary)
+        self.legacy_keyring = self._configure_legacy_backend()
+
+        KeyringWrapper.__shared_instance = self
+
+    def _configure_backend(self) -> Union[Any, FileKeyring]:
+        from chia.util.keychain import supports_keyring_passphrase
+
+        if self.keyring:
+            raise Exception("KeyringWrapper has already been instantiated")
+
+        if platform == "win32" or platform == "cygwin":
+            import keyring.backends.Windows
+
+            keyring.set_keyring(keyring.backends.Windows.WinVaultKeyring())
+        elif platform == "darwin":
+            if supports_keyring_passphrase():
+                keyring = FileKeyring(keys_root_path=self.keys_root_path)  # type: ignore
+            else:
+                import keyring.backends.macOS
```

Comment:
What is the rational between using FileKeyring before macOS backend?

Reply:
Lipa's been working on the GUI for passphrase support and prefers to work on macOS. This branch is never taken at the moment, but I can remove it to keep the code tidier until we're ready to roll it out more broadly.

Reply:
I was just thinking that the keyring backend would be more secure and therefore preferrable for storing keys. (hard to brute force etc, due to apple hardware defenses).